src/Combinatorics/Matroids/matroids.jl

```diff
@@ -0,0 +1,769 @@
+
+@doc Markdown.doc"""
+The cycle matroid of a graph.
+
+See Section 1.1 of Oxl11 (@cite)
+
+# Examples
+To construct the cycle matroid of the complete graph of 4 vertices write:
+```jldoctest
+julia> g = Oscar.Graphs.complete_graph(4);
+
+julia> M = cycle_matroid(g)
+Matroid of rank 3 on 6 elements
+```
+"""
+function cycle_matroid(g::Oscar.Graphs.Graph)
+    pm_Graph = Polymake.graph.Graph(ADJACENCY=g.pm_graph)
+    M = Polymake.matroid.matroid_from_graph(pm_Graph)
+    n = Oscar.Graphs.ne(g)
+    gs2num = Dict{Any,IntegerUnion}()
+    i = 1
+    for elem in 1:n
+        gs2num[elem] = i
+        i+=1
+    end
+    return Matroid(M,1:n,gs2num)
+end
+
+@doc Markdown.doc"""
+The `bond matroid` or `cocylce matroid` of a graph which is the dual of a cycle matroid, e.g, cographic.
+
+See Section 2.3 of Oxl11 (@cite)
+
+# Examples
+To construct the bond or cocycle matroid of the complete graph of 4 vertices write:
+```jldoctest
+julia> g = Oscar.Graphs.complete_graph(4);
+
+julia> M = bond_matroid(g)
+Matroid of rank 3 on 6 elements
+```
+
+or equivalently
+
+```jldoctest
+julia> g = Oscar.Graphs.complete_graph(4);
+
+julia> M = cocycle_matroid(g)
+Matroid of rank 3 on 6 elements
+```
+"""
+bond_matroid(g::Oscar.Graphs.Graph) = dual_matroid(cycle_matroid(g))
+
+@doc Markdown.doc"""
+See bond_matroid
+"""
+cocycle_matroid(g::Oscar.Graphs.Graph) = bond_matroid(g::Oscar.Graphs.Graph)
+
+
+@doc Markdown.doc"""
+The `dual matroid` of a given matroid.
+
+See page 65 and Sectrion 2 in Oxl11 (@cite)
+
+# Examples
+To construct the dual of the fano matroid write:
+```jldoctest
+julia> M = dual_matroid(fano_matroid())
+Matroid of rank 4 on 7 elements
+```
+"""
+dual_matroid(M::Matroid) = Matroid(M.pm_matroid.DUAL,M.groundset,M.gs2num)
+
+@doc Markdown.doc"""
+The ground set `E` of a matroid.
+
+To obtain the ground set of the fano matroid type:
+# Example
+```jldoctest
+julia> groundset(fano_matroid())
+7-element Vector{Int64}:
+ 1
+ 2
+ 3
+ 4
+ 5
+ 6
+ 7
+```
+"""
+groundset(M::Matroid) = M.groundset
+
+@doc Markdown.doc"""
+The `direct sum` of matroids.
+
+See Section 4.2 of Oxl11 (@cite)
+
+To obtain the direct sum of the fano and a uniform matroid type:
+# Example
+```jldoctest
+julia> direct_sum(fano_matroid(), uniform_matroid(2,4))
+Matroid of rank 5 on 11 elements
+```
+
+To take the sum of three uniform matroids use:
+# Example
+```jldoctest
+julia> matroids = Vector([uniform_matroid(2,4), uniform_matroid(1,3), uniform_matroid(3,4)]);
+
+julia> M = direct_sum(matroids)
+Matroid of rank 6 on 11 elements
+```
+"""
+function direct_sum(M::Matroid, N::Matroid)
+    gsN = N.groundset
+    while any(in(M.groundset), gsN)
+        gsN = Vector{Any}(copy(gsN))
+        for i in 1:length(gsN)
+            gsN[i] = string(gsN[i],'\'')
+        end
+    end
+    new_gs2num = Dict{Any,IntegerUnion}()
+    i = length(M.groundset)+1
+    for elem in gsN
+        new_gs2num[elem] = i
+        i+=1
+    end
```

Comment:
These four lines appear almost identically at least five times (also look at the other files). Please refactor to eliminate duplication.